sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryParallelReadTableSource.java

```diff
@@ -0,0 +1,216 @@
+                                BigQueryServices bqServices,
+                                ReadSessionOptions readOptions) {
+    this.bqServices = checkNotNull(bqServices, "bqServices");
+    this.coder = checkNotNull(coder, "coder");
+    this.parseFn = checkNotNull(parseFn, "parseFn");
+    this.jsonTable = checkNotNull(jsonTable, "jsonTable");
+    this.session = session;
+    this.tableSizeBytes = new AtomicReference<>();
+    this.initialReadLocation = readLocation;
+    this.readOptions = readOptions;
+  }
+
+  /**
+   * This method creates a {@link BigQueryParallelReadTableSource} with no initial session or read
+   * location.
+   */
+  public static <T> BigQueryParallelReadTableSource<T> create(
+      ValueProvider<TableReference> table,
+      SerializableFunction<SchemaAndRowProto, T> parseFn,
+      Coder<T> coder,
+      BigQueryServices bqServices,
+      ReadSessionOptions readOptions) {
+    return new BigQueryParallelReadTableSource<>(
+        NestedValueProvider.of(checkNotNull(table, "table"), new TableRefToJson()),
+        null,
+        // First TableSource shouldn't be used to CreateReader.
+        ReadLocation.getDefaultInstance(),
+        parseFn,
+        coder,
+        checkNotNull(bqServices),
+        readOptions);
+  }
+
+  /*
+   * Gets the total size of the table.
+   */
+  @Override
+  public synchronized long getEstimatedSizeBytes(PipelineOptions options) throws Exception {
+    if (tableSizeBytes.get() == null) {
+      TableReference table =
+          BigQueryIO.JSON_FACTORY.fromString(jsonTable.get(), TableReference.class);
+
+      Table tableRef = bqServices.getDatasetService(options.as(BigQueryOptions.class))
+          .getTable(table);
+      Long numBytes = tableRef.getNumBytes();
+      if (tableRef.getStreamingBuffer() != null) {
+        numBytes += tableRef.getStreamingBuffer().getEstimatedBytes().longValue();
+      }
+
+      tableSizeBytes.compareAndSet(null, numBytes);
+    }
+    return tableSizeBytes.get();
+  }
+
+  private static com.google.cloud.bigquery.v3.TableReferenceProto.TableReference
+  convertTableReferenceToV3(TableReference tableReference) {
+    com.google.cloud.bigquery.v3.TableReferenceProto.TableReference.Builder ref =
+        com.google.cloud.bigquery.v3.TableReferenceProto.TableReference.newBuilder();
+    return ref.setProjectId(tableReference.getProjectId())
+        .setDatasetId(tableReference.getDatasetId())
+        .setTableId(tableReference.getTableId()).build();
+  }
+
+  /**
+   * Creates a read session with the number of readers based on size estimation. This is
+   * only for the initial split when job is created. If the caller wants to split more, they will
+   * call Split function on the reader.
+   */
+  @Override
+  public List<BoundedSource<T>> split(
+      long desiredBundleSizeBytes, PipelineOptions options) throws Exception {
+    BigQueryOptions bqOptions = options.as(BigQueryOptions.class);
+    if (cachedSplitResult == null) {
```

Comment:
Does this source need such caching? The original one had a compelling reason: splitting is very expensive and can be called multiple times https://github.com/apache/beam/pull/2594/files#diff-f2e4daa48f5780a751295fff6b770950R88 . In this case,  does calling split multiple times also have disastrously expensive consequences?

Reply:
In our case, split() is where a read session is created; calling split() multiple times will result in orphaned read sessions, and while this isn't terrible -- they have a 24-hour TTL and self-delete after -- there's no user benefit to repeated split() calls.

Reply:
Are these split() calls done on the initial source, or on a source returned from the initial split() call, or on the reader? Note that we only create a new read session in the first case.